### DISCUSSION TITLE
better network security and performance via UNIX socket and SSH

### DISCUSSION THREAD

**Discussion by kolAflash** [distcc/distcc#517] [Ideas]
I was unhappy with the network security situation.  
https://www.distcc.org/security.html  
It seems I can only choose between no network security or bad performance. Especially, because in SSH mode every call to the compiler starts a new SSH process, going through the full SSH connection build up.

So I started thinking about it. And I'd be happy about your feedback :-)

&nbsp;

security.html says, that you should use a secure network. But nowadays I think that really needs to be an isolated network, which is not very practical when you like to use of all your machines standing around for compiling.
At least for my office network I can't assume It's secure. It allows wireless connections for guests coming to the office and there are all sorts of creepy devices in the network. And more importantly I also work on the machines I use for compiling. And when I run a web browser on these machines it allows JavaScript from foreign websites to make certain connections inside the network. Yes, the web browsers have a strange security model, but that's the way it is. There's an addon for Firefox and Chrome forbidding connections to local IP addresses, [but it can't detect if someone registers a public domainname to 127.0.0.1](https://translate.google.de/?&q=https://github.com/garywill/LAN-port-scan-forbidder/issues/1&sl=auto&tl=en&text=https%3A%2F%2Fgithub.com%2Fgarywill%2FLAN-port-scan-forbidder%2Fissues%2F1&op=translate). The only hope for web browsers is, that the browser vendors some day [might](https://wicg.github.io/private-network-access/) [close](https://developer.chrome.com/blog/private-network-access-update?hl=en) [that](https://bugzilla.mozilla.org/show_bug.cgi?id=354493) [gap](https://bugzilla.mozilla.org/show_bug.cgi?id=1481298).

So I started thinking if I could easily build a secure and efficient solution in another way. Two tools which came in handy where these:  
https://github.com/kohlschutter/unsock  
https://cwrap.org/socket_wrapper.html  
For the moment I choose to use Unsock. But Socket-Wrapper works pretty similar.  
With Unwrap you can make a program like Distcc transparently use a [UNIX socket](https://en.wikipedia.org/wiki/Unix_domain_socket) instead of a TCP socket.  
_(alternatively a feature request for Distcc would be an option to support UNIX sockets natively)_

&nbsp;

The UNIX socket can then be securely forwarded via SSH. Unfortunately SSH is single threaded. So it's worth making one SSH connection per CPU core on the worker machine. (you may make `cpu cores * 1.5` connections for an optimal workload)  
For herding this herd of SSH processes I use another socket on the worker to later shut them down together. So lets first open that "herding socket".  
worker (keep this process open while compiling):  
`socat STDIO UNIX-LISTEN:/home/USER/sockets/keepalive.sock,fork`  
Then start the SSH herd for lets say 16 cores:  
client:  
`for i in {10001..10016}; do ssh -f -L ~/sockets/"${i}".sock:"${HOME}"/sockets/3632.sock SSH_HOST socat STDOUT UNIX:/home/USER/sockets/keepalive.sock; done`  
These 16 SSH connections will then be used to connect to the worker. One for each compiler call running in parallel. The Socat command the SSH processes start will exit when you shut down the single Socat process started before. So you can shut down this herd of SSH processes altogether without imprecise tools like pkill. You can check for the SSH processes by running: `ps -wwfC ssh`

&nbsp;

Now let's start Distcc with Unwrap.
worker:  
`UNSOCK_ADDR=127.0.0.42/32 UNSOCK_DIR="${HOME}"/sockets LD_PRELOAD="${HOME}"/opt/unsock/libunsock.so.1.1.0 distccd -j$(nproc) --listen 127.0.0.42 --allow-private --daemon --no-detach`
client (assuming 8 cpu cores on the client and 16 on the worker):  
`UNSOCK_ADDR=127.0.0.42/32 UNSOCK_DIR="${HOME}"/sockets LD_PRELOAD="${HOME}"/opt/unsock/libunsock.so.1.1.0 DISTCC_HOSTS="localhost/8 $(for i in {10001..10016}; do echo -n "127.0.0.42:$i/1 "; done)" COMPILER_COMMAND`

&nbsp;

&nbsp;

Another possibility without UNIX sockets would be to use TCP with a localhost address (`--listen 127.0.0.1`) and to tunnel that TCP port to the other host via SSH. But in that case you'll also have a problem with the single threaded SSH. So you also need to start SSH multiple times and configure the distcc client with the separate socket addresses by these SSH processes.  
To secure this against locally running JavaScript inside web browsers you can restrict the TCP port to the users running distcc (that user then shouldn't run a web browser).  
```bash
iptables -A OUTPUT -p tcp --dport 8080 -m owner --uid-owner distccuser -j ACCEPT
iptables -A OUTPUT -p tcp --dport 8080 --destination 127.0.0.0/8 -j DROP
```

**Comment by sourcefrog** [comment]
I agree that plain TCP authenticated by IP isn't up to scratch today.

> Especially, because in SSH mode every call to the compiler starts a new SSH process, going through the full SSH connection build up.

This should not be true if you set SSH to reuse connections.

> The UNIX socket can then be securely forwarded via SSH. Unfortunately SSH is single threaded. So it's worth making one SSH connection per CPU core on the worker machine.

I don't think I follow the logic why you would need many SSH processes. My impression is that SSH can do reasonably well at forwarding multiple concurrently active connections.

Overall, it would be interesting to see performance numbers for this approach vs SSH with connection multiplexing (https://oooops.dev/2021/01/31/ssh-multiplexing-and-master-mode/).

**Comment by kolAflash** [comment]
> > Especially, because in SSH mode every call to the compiler starts a new SSH process, going through the full SSH connection build up.
> 
> This should not be true if you set SSH to reuse connections.

Interesting! Thanks for the hint. I've never heard of SSH connection reusing before. I found something called [ControlMaster](https://en.wikibooks.org/wiki/OpenSSH/Cookbook/Multiplexing). Is it what you're referring to?

https://manpages.debian.org/bullseye/openssh-client/ssh_config.5.en.html#ControlMaster
_Enables the sharing of multiple sessions over a single network connection._

So the SSH startup is faster. But the subsequent data transfer will always go through the one master thread which holds the connection. I would guess, that this is still considerably slower than one connection per worker cpu core.  
_(in theory you could start and use one SSH master process per worker cpu core, but I'd consider my above plan to be more straightforward)_

&nbsp;

> > The UNIX socket can then be securely forwarded via SSH. Unfortunately SSH is single threaded. So it's worth making one SSH connection per CPU core on the worker machine.
> 
> I don't think I follow the logic why you would need many SSH processes. My impression is that SSH can do reasonably well at forwarding multiple concurrently active connections.

With a single SSH connection I couldn't get the worker over 50 % cpu usage. And `ps -wwfLC ssh` shows, that there's only one thread in the SSH process. From my threading knowledge I'm pretty sure, that you can't get optimal performance with that if multiple compiler processes are pushing data through a single SSH thread.

&nbsp;

> Overall, it would be interesting to see performance numbers for this approach vs SSH with connection multiplexing (https://oooops.dev/2021/01/31/ssh-multiplexing-and-master-mode/).

For the moment I've to dig into my work again. But if someone likes to make some performance tests I'd be happy to see which solution actually has how much performance impact.

**Reply by sourcefrog** [reply]
Oh, interesting, I suppose if the client has many cores and is pretty unconstrained by running the preprocessor or other work, then SSH using a single core for crypto might well be the limiting factor. The fact that the client is at 50% CPU usage doesn't exactly prove this, maybe it's limited by remote throughput, but it's indicative.

**Comment by sourcefrog** [comment]
I think the long term solution would be something like TLS with some kind of mutual identity provider, natively in the program.
